src/nft/hooks/useFetchAssets.ts

```diff
@@ -0,0 +1,238 @@
+import { BigNumber } from '@ethersproject/bignumber'
+import { useWeb3React } from '@web3-react/core'
+import { GqlRoutingVariant, useGqlRoutingFlag } from 'featureFlags/flags/gqlRouting'
+import { useNftRouteLazyQuery } from 'graphql/data/__generated__/types-and-hooks'
+import { fetchRoute } from 'nft/queries'
+import { BagItemStatus, BagStatus, RouteResponse } from 'nft/types'
+import {
+  buildNftTradeInputFromBagItems,
+  buildSellObject,
+  recalculateBagUsingPooledAssets,
+  sortUpdatedAssets,
+} from 'nft/utils'
+import { buildRouteResponse } from 'nft/utils/nftRoute'
+import { combineBuyItemsWithTxRoute } from 'nft/utils/txRoute/combineItemsWithTxRoute'
+import { useCallback, useMemo } from 'react'
+import { useQueryClient } from 'react-query'
+import shallow from 'zustand/shallow'
+
+import { useBag } from './useBag'
+import { useSendTransaction } from './useSendTransaction'
+import { useTokenInput } from './useTokenInput'
+import { useTransactionResponse } from './useTransactionResponse'
+
+export function useFetchAssets(): () => Promise<void> {
+  const { account, provider } = useWeb3React()
+  const usingGqlRouting = useGqlRoutingFlag() === GqlRoutingVariant.Enabled
+  const sendTransaction = useSendTransaction((state) => state.sendTransaction)
+  const setTransactionResponse = useTransactionResponse((state) => state.setTransactionResponse)
+
+  const {
+    itemsInBag: uncheckedItemsInBag,
+    setBagStatus,
+    didOpenUnavailableAssets,
+    setDidOpenUnavailableAssets,
+    isLocked: bagIsLocked,
+    setLocked: setBagLocked,
+    setItemsInBag,
+    setBagExpanded,
+    reset: resetBag,
+  } = useBag(
+    ({
+      itemsInBag,
+      setBagStatus,
+      didOpenUnavailableAssets,
+      setDidOpenUnavailableAssets,
+      isLocked,
+      setLocked,
+      setItemsInBag,
+      setBagExpanded,
+      reset,
+    }) => ({
+      itemsInBag,
+      setBagStatus,
+      didOpenUnavailableAssets,
+      setDidOpenUnavailableAssets,
+      isLocked,
+      setLocked,
+      setItemsInBag,
+      setBagExpanded,
+      reset,
+    }),
+    shallow
+  )
+  const tokenTradeInput = useTokenInput((state) => state.tokenTradeInput)
+  const itemsInBag = useMemo(() => recalculateBagUsingPooledAssets(uncheckedItemsInBag), [uncheckedItemsInBag])
+
+  const queryClient = useQueryClient()
+  const [fetchGqlRoute] = useNftRouteLazyQuery()
+
+  const purchaseAssets = useCallback(
+    async (routingData: RouteResponse, purchasingWithErc20: boolean) => {
+      if (!provider || !routingData) return
+      const purchaseResponse = await sendTransaction(
+        provider?.getSigner(),
+        itemsInBag.filter((item) => item.status !== BagItemStatus.UNAVAILABLE).map((item) => item.asset),
+        routingData,
+        purchasingWithErc20
+      )
+
+      setBagLocked(false)
+      setTransactionResponse(purchaseResponse)
+      setBagExpanded({ bagExpanded: false })
+      resetBag()
```

Comment:
there was an if statement around this logic previously and I wonder if it's related to the two bugs I posted:
```
    if (
      purchaseResponse &&
      (transactionStateRef.current === TxStateType.Success || transactionStateRef.current === TxStateType.Failed)
    ) {

Reply:
yeah it was